### PR TITLE
qspi: stm32: Enhance the QSPI stm32 driver

### PR DESCRIPTION
**Pull request by lmajewski**
This pull request provides following enhancements to STM32 QSPI driver:
- Generation of the reset pulse for SPI-NOR memory
- Usage of 4IO for read / write (4READ/4PP)
- Support for different QSPI banks - bank1 and bank2 switch supported
- Support for 4B addressing on spi-nor

Not supported:
- The DMA operation (only interrupt mode supported)

Test HW:
- `nucleo_h743zi` with cape board

Test:
- `west build -p always -b nucleo_h743zi ./zephyr/samples/subsys/fs/littlefs`
- `west debug --gdb /home/lukma/.local/opt/zephyr-sdk-0.13.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gdb-no-py`

### CODE REVIEW

File: dts/bindings/flash_controller/st,stm32-qspi-nor.yaml
Comment (erwango) [inline dts/bindings/flash_controller/st,stm32-qspi-nor.yaml:49]: Let's add some info on the expected/supported values

File: dts/bindings/qspi/st,stm32-qspi.yaml
Comment (erwango) [inline dts/bindings/qspi/st,stm32-qspi.yaml:56]: required: ?
Also, please consider the use of default

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this great and clean addition.
First quick round of review

File: dts/bindings/qspi/st,stm32-qspi.yaml
Comment (lmajewski) [reply dts/bindings/qspi/st,stm32-qspi.yaml:56]: I think that the "default" and "required" properties are not needed. This option shall be used when somebody by mistake connect `quadspi_bk2` pins (which are dedicated to connecting second flash in two flash setup) to first flash.

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:883]: Ok, I will follow the used pattern.

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:772]: Ok.

File: dts/bindings/qspi/st,stm32-qspi.yaml
Comment (erwango) [reply dts/bindings/qspi/st,stm32-qspi.yaml:56]: Please add `required: false`

Comment (lmajewski) [reply dts/bindings/qspi/st,stm32-qspi.yaml:56]: Ok

**Comment by lmajewski** [comment]
@erwango Do you have any more comments?

File: samples/subsys/fs/littlefs/boards/nucleo_h743zi.conf
Comment (erwango) [inline samples/subsys/fs/littlefs/boards/nucleo_h743zi.conf:16]: Nit: Group the CONFIG_FS_LITTLEFS_foo together

**Review by erwango — CHANGES_REQUESTED** [review]
Minor changes requested, otherwise looks good.
Haven't been able to test on H7 but no regression on L4

**Review by erwango — APPROVED** [review]
LGTM.

@lmajewski If, interesredt, you can add yourself in CODEOWNER file in front of this PR. This could be done in a follow up PR.

**Comment by lmajewski** [comment]
@erwango I will prepare proper patch as a follow up. No problem. Thanks for the review :-)

**Comment by lmajewski** [comment]
> LGTM.
> 

Are there any other comments regarding this PR? Is it eligible for being pulled?

**Comment by erwango** [comment]
> Is it eligible for being pulled?

@lmajewski, to be eligible for merge, it requires a second approval.
@FRASTM, @ABOSTM , can you review ?

**Review by nvlsianpu — APPROVED** [review]

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:29]: The `DT_INST_NODE_HAS_PROP(0, reset_gpios)` is used so often that maybe it deserves some specific define, that would clarify the intention.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:709]: Shouldn't we adapt some errno ERR here? There is clearly intention for this to have some meaning because it gets printed in log here: https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R861

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:861]: I predict the error: it is -1.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [reply drivers/flash/flash_stm32_qspi.c:709]: Also in line https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R696 error gets returned from other function, `qspi_send_cmd`, which actually returns negative errno code. So now you have this function either return errno code or hardcoded -1.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:801]: Again errno? Fortunately log is, at least, giving us point of failure.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:878]: The function can fail and report that by returning in band error codes, so shouldn't this be checked here? Before doing anything else below?

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:708]: Since you use `if (ret)` in other places to test for non-0 ret, I suppose that this should be `!ret`, the same way you do here https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R754

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:880]: You can have success info even if the `qspi_program_quad_id` failed, and the `aspi_program_quad_id` has similar log at the end `https://github.com/zephyrproject-rtos/zephyr/pull/40859/files#diff-39156d185a094793a5d212e3ee4bbc9d275cf3baef6c5eca6aa95089521d1821R804` making this line redundant.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:875]: If `STM32_QSPI_QUAD_IO` is not enabled, the call seems redundant as there is not further processing of information gathered by call, nor the `rc` is used in any way.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:923]: It does not seem that `flash_stm32_qspi_gpio_reset` can ever fail, sot this code is redundant.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:867]: Is this still relevant if `STM32_QSPI_QUAD_IO` is not enabled?

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:875]: +1

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:878]: +1

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:880]: There is a difference between those two `LOG_INF()` information.

The one with "Mode: 1-4-4..." informs about the instruction code supported by the flash memory.
The one from the pasted above link informs that in SPI-NOR memory the QSPI mode has been enabled (the status register has correct bit set).

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:923]: +1

**Comment by erwango** [comment]
@lmajewski I'm setting label DNM to avoid early merge, as I understood you were about to push some of the comments reported by @nvlsianpu. Let me know when that's good for you.

**Comment by lmajewski** [comment]
> @lmajewski I'm setting label DNM to avoid early merge, as I understood you were about to push some of the comments reported by @nvlsianpu. Let me know when that's good for you.

Yes, I'm preparing new PR with changes proposed by @de-nordic

**Comment by erwango** [comment]
> Yes, I'm preparing new PR with changes proposed by @de-nordic

(I mixed the reviewers names, my bad)
Ok, you could also have fixed current PR, but since that's minor stuff, let's proceed that way.

**Comment by lmajewski** [comment]
@erwango @de-nordic 
I've submitted the new PR.

Most notable changes:
- Add CODEOWNERS entry
- Fix for 4READ command's read latency bits (correct naming and number for my flash)
- Fix the QE bit setting (it is a non-volatile bit, so need to be set only once).
- Change the code according to @de-nordic comments

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:163]: I think that the `IS_ENABLED` macro, although will work here, is intended to be used with `CONFIG_` definitions.

**Comment by erwango** [comment]
> I've submitted the new PR.

Thanks!
Just a quick clarification (so that we're all clear on what is happening):
What you've done is to push an update of current PR (which is fine and is indeed what was expected).
When you mention "I've submitted the new PR.", this is usually understood as "I have pushed a new branch and a new Github pull request has been created".
So no harm, everything fine, just a small wording issue :-).

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:25]: Redefine it as something `_ENABLED` because the `STM32_QSPI_QUAD_IO` sounds like index of some IO or number of IOs.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:784]: Just cosmetic: you may want remove empty line at 783 and add one above 782; the bit set is rather prerequisite for the `qspi_write_status_register` than post operation of the `qspi_write_enable`. It is a little bit dangling there.

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [inline drivers/flash/flash_stm32_qspi.c:864]: We can have 100 characters per line now, you do not have to use it everywhere but do not leave the lonely rc dangling in next line, it looks sad.

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:163]: I would prefer to leave it here.

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:25]: Maybe it shall be `STM32_QSPI_USE_QUAD_IO` ? 
(The same approach in the same file is with `STM32_QSPI_USE_DMA`)

I would like to avoid the situation when we would have `IS_ENABLED(STM32_QSPI_QUAD_IO_ENABLED)` as it looks
like some kind of the overkill.

Comment (de-nordic) [reply drivers/flash/flash_stm32_qspi.c:25]: Checkpatch failed to detect your usage if `IS_ENABLED` without config, this should generate warning: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch.pl#L6511-L6515

File: drivers/flash/flash_stm32_qspi.c
Comment (de-nordic) [reply drivers/flash/flash_stm32_qspi.c:163]: The checkpatch is shy to inform you that it would not like that: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch.pl#L6511-L6515

Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:163]: Ok, Then I will use:
`if (STM32_QSPI_QUAD_IO && dev_data->flag_quad_io_en)` directly in the if clause.

File: drivers/flash/flash_stm32_qspi.c
Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:25]: Ok, Then I will use `STM32_QSPI_USE_QUAD_IO` directly in the `if (` statement.

Comment (erwango) [reply drivers/flash/flash_stm32_qspi.c:25]: > Checkpatch failed to detect your usage if `IS_ENABLED` without config, this should generate warning: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/checkpatch.pl#L6511-L6515

This is not the case anymore, cf https://github.com/zephyrproject-rtos/zephyr/pull/39390

Comment (de-nordic) [reply drivers/flash/flash_stm32_qspi.c:25]: touché

Comment (lmajewski) [reply drivers/flash/flash_stm32_qspi.c:25]: So then I shall use:
`if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO) ... )` ?

Comment (de-nordic) [reply drivers/flash/flash_stm32_qspi.c:25]: Yes, you can use it.

**Pull request merged**
